### PR TITLE
Ensure datetime/date are serialized to correct format

### PR DESCRIPTION
**Pull request by nikclayton**
The `datetime` type should be emitted in RFC3339 format to match https://docs.joinmastodon.org/api/datetime-format/.

Ensure that happens by adding an `ActiveSupport::TimeWithZone` implementation with a `as_json` method that returns the value using `rfc3339(3)`.

The `date` type should be emitted in 'yyyy-mm-dd' format.

Add tests to ensure `datetime` and `date` are serialized correctly.

Since the existing tests don't hardcode the time it was simplest to make sure the `datetime` serialized value (which should be in RFC3339 format) can roundtrip through an RFC3339 parser.

### CODE REVIEW

**Comment by nikclayton** [comment]
Assume my first Ruby code was written about an hour ago, so there may well be better/more idiomatic ways to do this.

You can verify the test fails by modifying `time_with_zone.rb` and changing `rfc3339(3)` to `to_i`

Is the test appropriate? Verifying it round tripped seemed to be the simplest thing to do.

How many more tests are needed? I think it makes sense to duplicate the test for every field that should be serialized as RFC3339 datetime, but there's not a lot of existing tests so this may overwhelm them. On the other hand, got to start somewhere.

**Comment by renchap** [comment]
Related: https://github.com/mastodon/mastodon/pull/32704

We are considering changing the JSON library, and the new one has slightly different format for date.

**Comment by mjankowski** [comment]
Happy to do code review, but first as background ... are the motivations here...

- To fix some specific client interop or other breakage?
- To formalize something which should have been more precise already? ie - https://docs.joinmastodon.org/api/datetime-format/#datetime - does specify RFC3339
- Something else?

**Review by ClearlyClaire — APPROVED** [review]

**Review by ClearlyClaire — CHANGES_REQUESTED** [review]
There is actually a difference, which is that this code returns date and times for all objects, while we were sometimes returning only date on purpose.

**Comment by nikclayton** [comment]
@mjankowski 

> Happy to do code review,

Thanks

> but first as background ... are the motivations here...
> - To fix some specific client interop or other breakage?

Yes.

More accurately, saying a datetime is "ISO 8601" compatible is (after a bit of research) functionally useless information for interoperability. ISO 8601 is a complex standard that describes many different possible representations of dates and times (within the standard referred to as "complete", "truncated", "expanded", and more).

For example, only some ISO 8601 datetime combinations are required to specify a timezone indicator.

This causes real-world interoperability problems; for example, https://git.pleroma.social/pleroma/pleroma/-/issues/3261 (fixed with https://github.com/pachli/pachli-android/pull/1120).

The tests also prevent changes to the Mastodon code base from inadvertently changing part of the external API surface.

For example, in your PR #32704 the proposed change of output format from

>2024-11-13T14:47:02.846Z

to

> 2024-11-13 14:47:12 UTC

is a hard break for clients, as `2024-11-13 14:47:12 UTC` is not valid in any of the ISO8601 datetime representations. Specifically, the space between the date and time portion is invalid, as is the use of "UTC" as a time zone designator.

> To formalize something which should have been more precise already? ie - https://docs.joinmastodon.org/api/datetime-format/#datetime - does specify RFC3339

Yes. I wrote those docs after seeing this client breakage in the wild causing real problems for real users.

If these tests existed before #32704 they would also have caught the breakage ahead of time.

**Comment by ClearlyClaire** [comment]
> I think the other changes are good though -- featured tags returns a datetime (per the example in https://docs.joinmastodon.org/methods/featured_tags/#get, and I couldn't see that being clamped to a date in the code)

This sounds like the documentation is inaccurate or outdated, there's a `&.to_date` in the `app/serializers/rest/featured_tag_serializer.rb` and it does return a date only.

I believe there are also bare dates returned by `/api/v1/admin/measures`.

> and ditto for extended descriptions (https://docs.joinmastodon.org/entities/ExtendedDescription/).

Yes, that seems to be an error in the test itself, which creates a `ExtendedDescription` with a `Date` for `updated_at` whereas this would be a `DateTime` in practice.

> I could use some help writing a test in account_serializer_spec that verifies that `last_status_at` is returned as a date. The default account created in that test has a nil `last_status_at`, and my Ruby knowledge (now 90 minutes old) does not extend to figuring out how to create an instance of an account under test that has a non-nil value for that property.

This is a bit tricky because `last_status_at` is in a separate record that is supposed to be managed automatically when posting, but doing `account.account_stat.update!(last_status_at: your_value)` in a `before` block should do the trick.

**Comment by mjankowski** [comment]
Thanks for summary -- and yeah, as noted in that linked PR it's opened as a draft specifically because of this issue (to the extent there are specs, many of them are currently just doing a "does the value dumped match the value parsed" ... which is better than nothing and would catch any big datetime mistakes/changes, but is not really asserting anything about the format itself) and a few other things I wanted to nail down before actually merging that change.

It probably makes sense to:

- Add at least some coverage in the REST API request specs where we have important contract to uphold
- Add more exhaustive coverage at the serializer unit spec level (this currently barely exists, I did a first pass to get basic C0 coverage for stuff not hit by larger suite, needs expansion) ideally to cover every attribute, but at minimum to better coverage all attributes where we are doing any sort of even medium complexity logic/conversion/whatever. I have WIP on this but having been holding back to not overwhelm team. Will start opening these when they are available to review again.

**Comment by nikclayton** [comment]
@ClearlyClaire Thanks for that -- I'll check the docs/output (tomorrow at this rate) and update this PR accordingly.

@mjankowski Yeah -- this PR is the one that will extend the test suite to check the datetime and date info is serialized correctly.

**Comment by ClearlyClaire** [comment]
The tests do look good to me!

**Comment by nikclayton** [comment]
@mjankowski It's sort of done, but I need some pointers (or suggestions added directly to the PR, which is also fine).

1. If you search the diff for "pending" you'll find a handful of places where I'm not sure how to create the object in the necessary state to populate the relevant property.

2. Similarly, I couldn't see how to create an `Admin::Measure` to test the `data[][date]` property.

A couple of other things that are perhaps worth discussing, but can be done in different PRs (/cc @ClearlyClaire for feedback):

1. Should there be an equivalent PR that checks that **incoming** values to the API that should be string-date or string-datetime are of the expected shape? AIUI, the lowerlevel code currently does the equivalent of `DateTime.parse()` on the received string, and `DateTime.parse()` accepts more formats than just RFC3339.

   Including formats that don't specify a time, or a timezone.

   To me that feels like something that might be worth migrating to over a long time period (say, a year). Provide a server option to enable it with three settings, [allow, warn, deny]. `allow` would be current behaviour. `warn` would allow, but log details (e.g., the offending client and version, to allow outreach to developers of affected clients. And `deny` would reject requests with badly-formed dates.

2. These are "load bearing" tests. If they break the temptation might be to change the test, where the correct response is to either change the code back, or bump the API version number.

   Someone might not realise that though, so I wonder if a GitHub action that acts on PRs that change files under `spec/serializers/rest` and posts a canned comment warning about this possibility might be a good idea. It might be a little spammy, but the alternative -- inadvertent API version breakage -- could be worse.

**Comment by mjankowski** [comment]
Sounds good.

I'll do a code review pass with inline comments -- since they are all pretty similar I'll probably just do a few of them as I suspect any feedback will apply to multiple of them.

RE: the incoming values -- my take here would be as liberal as possible with incoming values - but only up to the point where there's not ambiguity (and raise errors only in those cases). But in either case, as you noted, out of scope for this PR. Can definitely follow up on that if there are specific edge cases to handle better.

RE: versioning or erroneous interpretation of breakage ... I think anything like that is true of basically *almost all of the api specs* and I'm not sure these are different. Presumably anyone authorized to merge code is competent enough to contemplate this possibility.

File: spec/serializers/rest/account_serializer_spec.rb
Comment (mjankowski) [inline spec/serializers/rest/account_serializer_spec.rb:6]: Probably `let(:default_datetime) { ...`

File: spec/serializers/rest/admin/account_serializer_spec.rb
Comment (mjankowski) [inline spec/serializers/rest/admin/account_serializer_spec.rb:9]: I havent looked at this one specifically to see if this is true or not - but we should prefer where possible to use the `Fabricate.build ...` form of these, which creates a valid instance of the model class, but does not persist to DB. I suspect for most of these serializers we should be able to just create a normal ruby object to pass in and shouldn't need actual DB records. In cases we do, fine to do so for now but might be interesting followup to see if we can make the serializers rely on passed in objects only and not need to query out while building.

Again I did not look at this one specifically, just general observation.

File: spec/serializers/rest/admin/cohort_serializer_spec.rb
Comment (mjankowski) [inline spec/serializers/rest/admin/cohort_serializer_spec.rb:18]: This is not a code review item per se, but I'd like to offer you a virtual handshake for your use of `datum` here. I hope it was truly enjoyable to do that.

File: spec/serializers/rest/admin/ip_serializer_spec.rb
Comment (mjankowski) [inline spec/serializers/rest/admin/ip_serializer_spec.rb:10]: RE: pending here - I suspect these are `UserIp` instances.

File: spec/fabricators/status_edit_fabricator.rb
Comment (nikclayton) [reply spec/fabricators/status_edit_fabricator.rb:5]: Doesn't `created_at` meet that criteria though? A `status_edit` without a `created_at` isn't valid.

File: spec/fabricators/ip_fabricator.rb
Comment (nikclayton) [reply spec/fabricators/ip_fabricator.rb:3]: Thanks. However, if I try this (`ip_fabricator.rb`):

```ruby
# frozen_string_literal: true

Fabricator(:user_ip) do
  user { Fabricate.build(:user) }
  ip { '10.0.0.1' }
  used_at { DateTime.new(2024, 11, 28, 16, 20, 0) }
end
```

I get this:

```
     Failure/Error: expect { DateTime.rfc3339(subject['used_at']) }.to_not raise_error

       expected no Exception, got #<ActiveRecord::ReadOnlyRecord: UserIp is marked as readonly> with backtrace:
         # /usr/share/rvm/gems/ruby-3.3.6/gems/fabrication-2.31.0/lib/fabrication/generator/base.rb:124:in `persist'
         # /usr/share/rvm/gems/ruby-3.3.6/gems/fabrication-2.31.0/lib/fabrication/generator/base.rb:31:in `create'
         # /usr/share/rvm/gems/ruby-3.3.6/gems/fabrication-2.31.0/lib/fabrication/schematic/definition.rb:86:in `block in fabricate'
         # /usr/share/rvm/gems/ruby-3.3.6/gems/fabrication-2.31.0/lib/fabrication/schematic/definition.rb:85:in `instance_eval'
         # /usr/share/rvm/gems/ruby-3.3.6/gems/fabrication-2.31.0/lib/fabrication/schematic/definition.rb:85:in `fabricate'
         # /usr/share/rvm/gems/ruby-3.3.6/gems/fabrication-2.31.0/lib/fabricate.rb:35:in `create'
         # /usr/share/rvm/gems/ruby-3.3.6/gems/fabrication-2.31.0/lib/fabrication.rb:66:in `Fabricate'
         # ./spec/serializers/rest/admin/ip_serializer_spec.rb:8:in `block (2 levels) in <top (required)>'
         # ./spec/serializers/rest/admin/ip_serializer_spec.rb:6:in `block (2 levels) in <top (required)>'
         # ./spec/serializers/rest/admin/ip_serializer_spec.rb:12:in `block (4 levels) in <top (required)>'
         # ./spec/serializers/rest/admin/ip_serializer_spec.rb:12:in `block (3 levels) in <top (required)>'
         # ./spec/rails_helper.rb:134:in `block (2 levels) in <top (required)>'
         # /usr/share/rvm/gems/ruby-3.3.6/gems/webmock-3.24.0/lib/webmock/rspec.rb:39:in `block (2 levels) in <top (required)>'
     # ./spec/serializers/rest/admin/ip_serializer_spec.rb:12:in `block (3 levels) in <top (required)>'
     # ./spec/rails_helper.rb:134:in `block (2 levels) in <top (required)>'
     # /usr/share/rvm/gems/ruby-3.3.6/gems/webmock-3.24.0/lib/webmock/rspec.rb:39:in `block (2 levels) in <top (required)>'
```

?

File: spec/serializers/rest/account_serializer_spec.rb
Comment (nikclayton) [reply spec/serializers/rest/account_serializer_spec.rb:6]: Done. Thanks.

File: spec/serializers/rest/admin/account_serializer_spec.rb
Comment (nikclayton) [reply spec/serializers/rest/admin/account_serializer_spec.rb:9]: I'm just following the existing code style (e.g., the usage in `status_serializer_spec.rb`, which looks like:

```ruby
  let(:current_user) { Fabricate(:user) }
  let(:alice) { Fabricate(:account, username: 'alice') }
  let(:bob)   { Fabricate(:account, username: 'bob', domain: 'other.com') }
  let(:status) { Fabricate(:status, account: alice) }
```

File: spec/fabricators/ip_fabricator.rb
Comment (mjankowski) [reply spec/fabricators/ip_fabricator.rb:3]: Ah, yes ... `UserIp` wraps a DB view, which populates rows from querying for user-related IP values from a few diff tables. For purposes of using it in a spec, something like...

`UserIp.new(used_at: 3.days.ago)` would probably work, and could skip fabricator.

Comment (nikclayton) [reply spec/fabricators/ip_fabricator.rb]: That works. Thanks.

File: spec/serializers/rest/admin/measure_serializer_spec.rb
Comment (mjankowski) [inline spec/serializers/rest/admin/measure_serializer_spec.rb:5]: For this one - the `retrieve`  method of `Measure` is going to return an array of measures, of varying class types. Its fine to leave the setup that way and assert about the first  instance or something, but you could also just used one of the classes, ie:

`let(:record) { Admin::Metrics::Measure::ActiveUsersMeasure.new(start_at, end_at, nil) }`

Have not double checked but I think they are set up that way.

Then the assertions might also need updating re: whatever the returned json format is here.

File: spec/fabricators/admin_report_fabricator.rb
Comment (mjankowski) [reply spec/fabricators/admin_report_fabricator.rb:3]: I think the new fabricator is unnecessary, yes -- both serializers take `Report` instances, so both could use the existing fabricator.

File: spec/serializers/rest/admin/measure_serializer_spec.rb
Comment (nikclayton) [reply spec/serializers/rest/admin/measure_serializer_spec.rb]: Thanks. Updated and fixed.

File: spec/fabricators/admin_report_fabricator.rb
Comment (nikclayton) [reply spec/fabricators/admin_report_fabricator.rb:3]: Thanks. Fixed.

File: spec/fabricators/status_edit_fabricator.rb
Comment (nikclayton) [reply spec/fabricators/status_edit_fabricator.rb]: Ack. Done.

File: spec/serializers/rest/notification_group_serializer_spec.rb
Comment (mjankowski) [inline spec/serializers/rest/notification_group_serializer_spec.rb:15]: For this one - there is a `NotificationGroup` class, but its not a db-backed AR model ... it is also a serializer. Even more nuanced, it relies on a `pagination_data` object to be present to even return the relevant datetime.

Bare minimum might be something like...

```
mastodon(dev)> ng = NotificationGroup.new pagination_data: { latest_notification_at: 3.days.ago }, notification: Fabricate(:notification), sample_accounts: []
mastodon(dev)> REST::NotificationGroupSerializer.new(ng).to_json
```

Comment (nikclayton) [reply spec/serializers/rest/notification_group_serializer_spec.rb]: That works. Thanks.

File: spec/serializers/rest/account_serializer_spec.rb
Comment (mjankowski) [inline spec/serializers/rest/account_serializer_spec.rb:69]: Is this one actually re: something in the `AccountSerializer` ... or should be in separate `MutedAccountSerializer` spec?

In either case, something like...

`Fabricate :mute, account: x, target_account: y, mute_expires_at: 2.days.ago` - where `x` is instance of the account that created the mute and `y` is instance of account it wants to mute, would create the background condition here.

File: spec/serializers/rest/account_serializer_spec.rb
Comment (mjankowski) [inline spec/serializers/rest/account_serializer_spec.rb:79]: These "verified" come from the "fields" on account. You can get a field with something like....

`Account::Field.new(account, 'name' => 'Foo', 'value' => 'Bar', 'verified_at' => 3.days.ago)`

...and then would need `Account#fields` to return an array w/ that in it.

File: spec/serializers/rest/status_serializer_spec.rb
Comment (mjankowski) [inline spec/serializers/rest/status_serializer_spec.rb:61]: There's an `UpdateStatusService` which is actually used to do this ... but you might be able to get away with just `Fabricate.build :status, edited_at: 3.days.ago` or something for purposes of this serialization spec.

File: spec/serializers/rest/account_serializer_spec.rb
Comment (ClearlyClaire) [inline spec/serializers/rest/account_serializer_spec.rb:74]:
```suggestion
      account.mute!(target_account)
```

File: spec/serializers/rest/account_relationship_severance_event_serializer_spec.rb
Comment (ClearlyClaire) [inline spec/serializers/rest/account_relationship_severance_event_serializer_spec.rb:8]:
```suggestion
  let(:record) { Fabricate.build :account_relationship_severance_event, id: 123, created_at: DateTime.new(2024, 11, 28, 16, 20, 0).utc }
```

**Review by ClearlyClaire — CHANGES_REQUESTED** [review]

File: spec/serializers/rest/account_serializer_spec.rb
Comment (nikclayton) [reply spec/serializers/rest/account_serializer_spec.rb]: Couldn't get that to work, so used the account's follow! and mute! methods.

File: spec/serializers/rest/account_relationship_severance_event_serializer_spec.rb
Comment (nikclayton) [reply spec/serializers/rest/account_relationship_severance_event_serializer_spec.rb:8]: Can you help me understand the rationale for that change?

As long as it's a serialized as an RFC3339 compatible date client's shouldn't mind what the timezone designator is. Even if https://docs.joinmastodon.org/api/datetime-format/ was updated to say the server will only ever return dates in UTC (so the timezone designator would always be `Z`) clients should be prepared for it not to be.

Also, I suspect it provides more opportunities for bugs where the server inadvertently returns a correctly formatted RFC3339 datetime with a timezone offset.

So I'm concerned that being explicit that it must be UTC in the tests is overspecifying the test relative to the real-world behaviour that's expected.

File: spec/serializers/rest/account_serializer_spec.rb
Comment (nikclayton) [reply spec/serializers/rest/account_serializer_spec.rb]: Thanks, that was the pointer I needed.

**Comment by nikclayton** [comment]
That's the last of the pending tests implemented, thanks @ClearlyClaire @mjankowski.

File: config/initializers/time_with_zone.rb
Comment (mjankowski) [inline config/initializers/time_with_zone.rb:6]: I know some are still pending and this is (maybe?) still somewhat WIP -- but I'm curious, are there specs here which fail when this is not present?

It's a bit of a maze to navigate because we don't always have the same starting classes (I think that AR model timestamp columns do produce `ActiveSupport::TimeWithZone`, but there may be other things using `Date`, `DateTime`, etc - or we may be overriding/casting values in place).

I mostly want to understand what this initializer is doing. When I look at the default implementation - https://github.com/rails/rails/blob/v7.2.2/activesupport/lib/active_support/time_with_zone.rb#L154-L172 - the behavior is contingent on a setting for standard json format which defaults to true and we dont override, and then also on the configured time precision, which defaults to 3.

Separately, within that class - https://github.com/rails/rails/blob/v7.2.2/activesupport/lib/active_support/time_with_zone.rb#L144-L152 - the `iso8601` and the `rfc3339` methods are both aliases for `xmlschema` which takes that precision argument and calls out to formatting methods to actually make the string.

So I guess, upon my naive reading here, it seems like the thing which the monkeypatch is doing ... is already the default behavior? Is there another override somewhere in the AMS layer or elsewhere that we are re-overriding here?

As a small example...

```
mastodon(test)> a = Time.zone.now
=> Fri, 06 Dec 2024 21:02:42.077544000 UTC +00:00
mastodon(test)> a.iso8601
=> "2024-12-06T21:02:42Z"
mastodon(test)> a.rfc3339
=> "2024-12-06T21:02:42Z"
mastodon(test)> a.as_json
=> "2024-12-06T21:02:42.077Z"
mastodon(test)> a.class
=> ActiveSupport::TimeWithZone
```

then converted...

```
mastodon(test)> b = a.to_datetime
=> Fri, 06 Dec 2024 21:02:42 +0000
mastodon(test)> b.iso8601
=> "2024-12-06T21:02:42+00:00"
mastodon(test)> b.rfc3339
=> "2024-12-06T21:02:42+00:00"
mastodon(test)> b.as_json
=> "2024-12-06T21:02:42.077+00:00"
```

So there is a difference there between the `AS::TimeWithZone` and `DateTime` json serializations ... but the default behavior of AS:TWZ seems not changed by this config?

**Comment by mjankowski** [comment]
> That's the last of the pending tests implemented, thanks @ClearlyClaire @mjankowski.

Great -- left one (sort of lengthy!) comment on the initializer ... with those pendings gone, is the rest ready for final review, or still some WIP in here?

File: config/initializers/time_with_zone.rb
Comment (nikclayton) [reply config/initializers/time_with_zone.rb:6]: I think you're correct that setting it here does not change the default behaviour.

However, setting it specifically in `time_with_zone.rb` does a few things:

1. It guards against any future changes to the Rails default.
2. It guards against any code that changes the value of `.time_precision`.
3. It clearly documents intentions.

I'll extend the comment to note that this is because of an external API commitment.

Comment (nikclayton) [reply config/initializers/time_with_zone.rb:9]: Also, to re-iterate, reading any code/comments that talk about "ISO 8601 format" should set off alarm bells that the author was being too vague and the code should be treated very carefully.

RFC3339 is a concrete, actually useful definition.

XMLSchema came after (as far as I can tell), and re-used most of the RFC3339 definition without crediting it (https://www.w3.org/TR/xmlschema-2/#dateTime). It's slightly stricter than RFC3339 (e.g., `T` as a separator rather than `T` or `t`).

**Comment by nikclayton** [comment]
> with those pendings gone, is the rest ready for final review, or still some WIP in here?

Ready for (the hopefully final) review. Thanks both for your very useful feedback so far.

File: config/initializers/time_with_zone.rb
Comment (mjankowski) [reply config/initializers/time_with_zone.rb:9]: Thanks for confirming I wasn't missing something!

I guess if we're worried about the framework defaults changing and trying to protect against that, my choices in rough order of preference would be:

- Make sure the serializer specs would start failing if any of the things we care about in the json date format break during a dependency update
- If for whatever reason we can't full capture that in those specs, or we're worried about a careless developer just blindly changing the specs to "fix" the failures -- add an explicit (non serializer) spec somewhere that has a very verbose setup/description/comments/etc to make clear why we care and is purely about generic json date format serialization, and do an exact string comparison to a known date value (instead of generic "does this parse" check)
- If we need to add a configuration setting to enable either of those prior items (ie, set the precision value, set the format style, whatever), do that in an initializer
- Only as a last resort fall back to monkey patch if we can't be confident with any of the prior

**Review by ClearlyClaire — CHANGES_REQUESTED** [review]

**Comment by mjankowski** [comment]
> Ready for (the hopefully final) review. Thanks both for your very useful feedback so far.

Other than the open comments about linting stuff and the monkey patch, I don't have much else. I think some future refactors here might be useful, but aren't necessary for this initial PR:

- Do a manual pass over serializers/spec to be sure we've captured every datetime field here
- Do pass over specs to see if there are places we can use PORO or `Fabricate.build` instead of persisted records
- If/when we add more serializer coverage that's not specific to the date values - think about pulling out the date check to a matcher or something so lines read like `expect(subject['whatever_at']).to be_serialized_as_api_datetime` (naming TBD), would be a nice DRY/readability touch, and more fully capture the "what do we care about in an api date format?" in one spot, called from misc serializer/request specs as needed.

File: spec/serializers/rest/account_relationship_severance_event_serializer_spec.rb
Comment (nikclayton) [reply spec/serializers/rest/account_relationship_severance_event_serializer_spec.rb:8]: Where did you see the spec failure?

If I run `rspec spec/serializers/rest/account_relationship_severance_event_serializer_spec.rb` in a clean client I get:

```
Run options: exclude {:streaming=>true, :search=>true, :js=>true}

Randomized with seed 19649
..

Finished in 1.98 seconds (files took 8.68 seconds to load)
2 examples, 0 failures

Randomized with seed 19649

Coverage report generated for RSpec to /home/nik/projects/mastodon/mastodon/coverage.
Line Coverage: 4.44% (1830 / 41211)
Branch Coverage: 3.77% (49 / 1300)
```

> The application code uses, or at least should use, UTC dates everywhere.

Two things here:

1. If I've understood Rails correctly (and I may not have), everything should be using whatever timezone Rails is configured to use (the default is UTC). So if you need to put `.utc` here to get a test to pass that suggests something is wrong somewhere else.

2. I don't think that should be relevant here. Specifically, I think that should be treated as an internal Mastodon decision. For the purposes of conformance and client interoperability (which is what the serializer specs test) it is sufficient that this is an RFC3339 datetime. Whether the timezone designator is `Z` or an offset like `+02:00` shouldn't matter. And  "Mastodon will only ever emit RFC3339 datetimes in UTC" seems like an unnecessarily strong external commitment to make.

File: config/initializers/time_with_zone.rb
Comment (ClearlyClaire) [reply config/initializers/time_with_zone.rb:9]: I still think this is best left out of this PR.

**Comment by mjankowski** [comment]
Thanks! - I think this is probably good to go as a first pass here. Can probably clean anything up in followups.

(Needs another CI approval, though...)

**Review by ClearlyClaire — APPROVED** [review]
Looks good to me now, thanks again for your contribution!

**Comment by mjankowski** [comment]
Excellent. Will rebase (and convert to ready, I think) the json gem one when this merges.

**Pull request merged**
